File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_449/java/Solution.java
package com.hi.dhl.algorithms.leetcode._449.java;

import com.hi.dhl.algorithms.model.ListNode;
/**
 * <pre>
 *     author: dhl
 *     date  : 2022/4/27
 *     desc  :
 * </pre>
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = null;
        l1 = reverse(l1);
        l2 = reverse(l2);
        int carry = 0;
        while (l1 != null || l2 != null) {
            int a = l1 == null ? 0 : l1.val;
            int b = l2 == null ? 0 : l2.val;
            int sum = a + b + carry;
            carry = sum / 10;
            ListNode node = new ListNode(sum % 10);
            node.next = head;
            head = node;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }

        if (carry != 0) {
            ListNode node = new ListNode(carry);
            node.next = head;
            head = node;
        }
        return head;
    }

    public ListNode reverse(ListNode head) {
        ListNode pre = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = pre;
            pre = head;
            head = next;
        }
        return pre;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_449/kotlin/Solution.kt
package com.hi.dhl.algorithms.leetcode._449.kotlin

import com.hi.dhl.algorithms.model.ListNode

/**
 * <pre>
 *     author: dhl
 *     date  : 2022/4/27
 *     desc  :
 * </pre>
 */
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var p1 = reverse(l1)
        var p2 = reverse(l2)
        var head: ListNode? = null;
        var carry = 0;
        while(p1 != null || p2 != null || carry > 0){
            val a = p1?.`val` ?:0
            val b = p2?.`val` ?:0
            val sum = a +b + carry
            carry = sum / 10;
            var node = ListNode(sum % 10)
            node.next = head
            head = node
            p1 = p1?.next
            p2 = p2?.next
        }
        return head
    }

    fun reverse(head:ListNode?): ListNode?{
        var cur = head
        var pre:ListNode? = null
        while(cur != null){
            var next = cur.next;
            cur.next = pre
            pre = cur
            cur = next
        }
        return pre
    }
}